src/detect-xor.c

```diff
@@ -0,0 +1,1080 @@
+                    key_len, UINT16_MAX);
+            goto cleanup;
+        } else if (key_len > 0) {
+            size_t tmp_len = 0;
+
+            data->key = BytesFromHexString(key_str, &tmp_len);
+            data->key_len = (uint16_t) tmp_len;
+
+            if (data->key == NULL || data->key_len == 0) {
+                SCLogError(SC_ERR_INVALID_RULE_ARGUMENT,
+                    "Bad value for xor key: \"%s\"", key_str);
+                goto cleanup;
+            }
+        }
+    }
+
+    if (pcre_rc >= 2) {
+        if (pcre_get_substring((char *)str, ov, max, 2, &key_be_str) > 0) {
+            SigMatch *bed_sm = DetectByteExtractRetrieveSMVar(key_be_str, s);
+            if (bed_sm == NULL) {
+                SCLogError(SC_ERR_INVALID_SIGNATURE,
+                        "Bad value for xor key: \"%s\". "
+                        "Unknown byte_extract variable. "
+                        "For literal hex value, surround with double quotes.\n",
+                        key_be_str);
+                goto cleanup;
+            }
+
+            DetectByteExtractData *bed = (DetectByteExtractData *) bed_sm->ctx;
+
+            data->bid = bed->local_id;
+
+            if (bed->flags & DETECT_BYTE_EXTRACT_FLAG_STRING) {
+                /* bytes were extracted from a string, can't use nbytes to
+                 * determine key length so use entire value */
+                data->key_len = 8;
+            } else {
+                /* bytes were extracted directly, use nbytes as key length */
+                data->key_len = bed->nbytes;
+            }
+
+            if (data->key_len == 0 || data->key_len > 8) {
+                SCLogError(SC_ERR_INVALID_SIGNATURE,
+                        "Bad value for xor key: \"%s\". "
+                        "byte_extract variable length should be between 1 and "
+                        "8 but was %"PRIu16".\n",
+                        key_str, data->key_len);
+                goto cleanup;
+            }
+
+            /** \todo improve cache coherence by embedding the key array inside
+             * the context struct instead of allocating it. This could work well
+             * for short key sizes (i.e. up to 8 bytes) and be represented as a
+             * union. This would require us to track when we choose to allocate
+             * using a flag. */
+            data->key = SCCalloc(data->key_len, sizeof(*data->key));
+
+            if (unlikely(data->key == NULL)) {
+                goto cleanup;
+            }
+        }
+    }
+
+    if (data->key == NULL && data->bid < 0) {
+        SCLogError(SC_ERR_INVALID_RULE_ARGUMENT,
+                "Bad value for xor key: \"%s\". "
+                "Expected hex string in double quotes or byte_extract variable.",
+                key_str);
+        goto cleanup;
+    }
+
+     if (pcre_rc >= 3) {
+        if (pcre_get_substring((char *)str, ov, max, 3, &bytes_str)) {
```

Comment:
in general we strive to use pcre_copy_substring over pcre_get_substring as the latter allocs on the heap.